lib/Analysis/Utility.cpp

```diff
@@ -73,6 +73,20 @@ unsigned ReduceOpHelper::getThreadOffsetOnReductionAxis() {
     auto parentLayout = sliceLayout.getParent();
     auto threadsPerWarp = getThreadsPerWarp(parentLayout);
     threadOffset = threadsPerWarp[sliceLayout.getDim()];
+    // For cases where there is another dimension in between sliceDim and axis:
+    auto sliceDim = sliceLayout.getDim();
+    auto order = getOrder(parentLayout);
+    int axisOrder = -1, sliceDimOrder = -1;
+    for (unsigned i = 0; i < order.size(); i++) {
+      if (order[i] == axis)
+        axisOrder = i;
+      if (order[i] == sliceDim)
+        sliceDimOrder = i;
+    }
+    assert(axisOrder >= 0 && sliceDimOrder >= 0);
+    // This loop does nothing if sliceDimOrder >= axisDimOrder.
+    for (unsigned i = sliceDimOrder + 1; i < axisOrder; i++)
+      threadOffset *= threadsPerWarp[order[i]];
```

Comment:
I don't think this is a fully general solution. It would not work with higher rank cases for instance? 
Could you try to generalize the logic?

Reply:
It should work with 4D or higher-dim tensors as it looks at the sliceDim and the axisDim for the reduction, without depending on the rank. I am not sure whether it is possible that sliceDimOrder can be higher than axisDimOrder or not, and if it is, what should happen.
For the case where sliceDimOrder < axisDimOrder, we are multiplying threadsPerWarp for orders in between.

Reply:
I meant for cases where there are multiple reductions chained and where the layout would be a sliced of a sliced of a sliced?

Reply:
@ThomasRaoux Does this example match what you meant?
```
  tt.func @reduce_3d(%arg0 : tensor<2x2xf32, #triton_gpu.slice<{dim = 2, parent = #blocked}>>) {
    %1 = "tt.reduce"(%arg0) <{axis = 1 : i32}> ({
        ^bb0(%arg1: f32, %arg2: f32):
          %112 = arith.addf %arg1, %arg2 : f32
          tt.reduce.return %112 : f32
        }) : (tensor<2x2xf32, #triton_gpu.slice<{dim = 2, parent = #blocked}>>) -> tensor<2xf32, #triton_gpu.slice<{dim = 1, parent = #triton_gpu.slice<{dim = 2, parent = #blocked}>}>>
    // the second one take as input a slice of slice, then output a slice of slice of slice
    %0 = "tt.reduce"(%1) <{axis = 0 : i32}> ({
        ^bb0(%arg3: f32, %arg4: f32):
          %113 = arith.addf %arg3, %arg4 : f32
          tt.reduce.return %113 : f32
        }) : (tensor<2xf32, #triton_gpu.slice<{dim = 1, parent = #triton_gpu.slice<{dim = 2, parent = #blocked}>}>>)
        -> f32
    tt.return
  }
  ```

Reply:
The test case makes sense to me. We can probably check against other implementations to make sure it exposes a runtime correctness issue.

Reply:
yes that's the kind of chained cases that would cause problem although for the original problem to show up we would probably need more dimensions to reproduce this problem multiple times:
```
    // For cases where there is another dimension in between sliceDim and axis:
```

Anyway I think you understood what I meant and I believe current code wouldn't handle this